src/snowflake/snowpark/dataframe.py

```diff
@@ -3364,6 +3365,78 @@ def describe(self, *cols: Union[str, List[str]]) -> "DataFrame":
         )
         return res_df
 
+    @df_api_usage
+    def rename(
+        self,
+        col_or_mapper: Union[ColumnOrName, dict],
+        new_column: str = None,
+    ):
+        """
+        Returns a DataFrame with the specified column ``col_or_mapper`` renamed as ``new_column``. If ``col_or_mapper``
+        is a dictionary, multiple columns will be renamed in the returned DataFrame.
+
+        Example::
+            >>> # This example renames the column `A` as `NEW_A` in the DataFrame.
+            >>> df = session.sql("select 1 as A, 2 as B")
+            >>> df_renamed = df.rename(col("A"), "NEW_A")
+            >>> df_renamed.show()
+            -----------------
+            |"NEW_A"  |"B"  |
+            -----------------
+            |1        |2    |
+            -----------------
+            <BLANKLINE>
+            >>> # This example renames the column `A` as `NEW_A` and `B` as `NEW_B` in the DataFrame.
+            >>> df = session.sql("select 1 as A, 2 as B")
+            >>> df_renamed = df.rename({col("A"): "NEW_A", "B":"NEW_B"})
+            >>> df_renamed.show()
+            ---------------------
+            |"NEW_A"  |"NEW_B"  |
+            ---------------------
+            |1        |2        |
+            ---------------------
+            <BLANKLINE>
+
+        Args:
+            col_or_mapper: The old column instance or column name to be renamed, or the dictionary mapping from column instances or columns names to their new names (string)
+            new_column: The new column name (string value), if a single old column is given
+        """
+        if new_column is not None:
+            return self.with_column_renamed(col_or_mapper, new_column)
+
+        if not isinstance(col_or_mapper, dict):
+            raise ValueError(
+                f"If new_column parameter is not specified, col_or_mapper needs to be of type dict, "
+                f"not {type(col_or_mapper).__name__}"
+            )
+
+        if len(col_or_mapper) == 0:
+            raise ValueError("col_or_mapper dictionary cannot be empty")
+
+        column_or_name_list, rename_list = zip(*col_or_mapper.items())
+        for name in rename_list:
+            if not isinstance(name, str):
+                raise ValueError(
+                    f"You cannot rename a column using value {name} of type {type(name).__name__} as it "
+                    f"is not a string."
+                )
+
+        names = self._get_column_names_from_column_or_name_list(column_or_name_list)
+        normalized_name_list = [quote_name(n) for n in names]
```

Comment:
you could use analyzer.analyze to convert `Column` expressions to respective `str`.

Also you could convert `str`s into `Column` objects using `_to_col_if_str` and use the resolver approach used for `Project` statement
https://github.com/snowflakedb/snowpark-python/blob/c6e9b569aa227713eba0aab8b6c0159315a53679/src/snowflake/snowpark/_internal/analyzer/analyzer.py#L798#L801

Reply:
created a JIRA to handle this separately: https://snowflakecomputing.atlassian.net/browse/SNOW-847788